File: fm-domain/src/main/java/io/zeitwert/fm/building/adapter/api/rest/ProjectionController.java
package io.zeitwert.fm.building.adapter.api.rest;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import io.zeitwert.fm.building.model.enums.CodeBuildingPart;
import io.zeitwert.fm.building.service.api.dto.ProjectionPeriod;

@RestController("projectionController")
@RequestMapping("/rest/projection")
public class ProjectionController {

	@GetMapping("/elements/{elementId}")
	ResponseEntity<List<ProjectionPeriod>> getElementProjection(@PathVariable String elementId)
			throws InterruptedException, ExecutionException {
		final CodeBuildingPart buildingPart = CodeBuildingPart.getBuildingPart(elementId);
		final int startYear = 2021;
		final double condition = 1;
		final int duration = 50;
		final double value = 1000000.0;
		List<ProjectionPeriod> projection = buildingPart.getProjection(value, startYear, condition, startYear, duration);
		return ResponseEntity.ok(projection);
	}

	@GetMapping("/elements/{elementId}/withRestoration")
	ResponseEntity<List<Double>> getRestoredTimeValue(@PathVariable String elementId)
			throws InterruptedException, ExecutionException {
		CodeBuildingPart buildingPart = CodeBuildingPart.getBuildingPart(elementId);
		List<Double> timeValues = new ArrayList<Double>();
		int year = 0;
		for (int t = 0; t < 100; t++) {
			Double timeValue = buildingPart.getTimeValue((double) year);
			timeValues.add(timeValue);
			if (timeValue <= buildingPart.getOptimalRestoreTimeValue()) {
				year = 0;
			}
			year += 1;
		}
		return ResponseEntity.ok(timeValues);
	}

	@GetMapping("/elements/{elementId}/relativeAge/{timeValue}")
	ResponseEntity<Double> getRelativeAge(@PathVariable String elementId, @PathVariable Double timeValue)
			throws InterruptedException, ExecutionException {
		CodeBuildingPart buildingPart = CodeBuildingPart.getBuildingPart(elementId);
		return ResponseEntity.ok(buildingPart.getRelativeAge(timeValue));
	}

	@GetMapping("/elements/{elementId}/nextRestoration/{ratingYear}/{condition}")
	ResponseEntity<ProjectionPeriod> getNextRestoration(@PathVariable String elementId,
			@PathVariable Integer ratingYear,
			@PathVariable Double condition)
			throws InterruptedException, ExecutionException {
		CodeBuildingPart buildingPart = CodeBuildingPart.getBuildingPart(elementId);
		return ResponseEntity
				.ok(buildingPart.getNextRestoration(1000000, ratingYear, condition));
	}

}

File: fm-domain/src/main/java/io/zeitwert/fm/building/adapter/rest/ProjectionController.kt
package io.zeitwert.fm.building.adapter.rest

import io.zeitwert.fm.building.api.dto.ProjectionPeriod
import io.zeitwert.fm.building.model.enums.CodeBuildingPart.Enumeration.getBuildingPart
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.concurrent.ExecutionException

@RestController("projectionController")
@RequestMapping("/rest/projection")
class ProjectionController {

	@GetMapping("/elements/{elementId}")
	@Throws(InterruptedException::class, ExecutionException::class)
	fun getElementProjection(
		@PathVariable elementId: String,
	): ResponseEntity<List<ProjectionPeriod>> {
		val buildingPart = getBuildingPart(elementId)!!
		val startYear = 2021
		val condition = 1.0
		val duration = 50
		val value = 1000000.0
		val projection = buildingPart.getProjection(value, startYear, condition, startYear, duration)
		return ResponseEntity.ok(projection)
	}

	@GetMapping("/elements/{elementId}/withRestoration")
	@Throws(InterruptedException::class, ExecutionException::class)
	fun getRestoredTimeValue(
		@PathVariable elementId: String,
	): ResponseEntity<List<Double>> {
		val buildingPart = getBuildingPart(elementId)!!
		val timeValues: MutableList<Double> = mutableListOf()
		var year = 0
		for (t in 0..99) {
			val timeValue = buildingPart.getTimeValue(year.toDouble())
			timeValues.add(timeValue)
			if (timeValue <= buildingPart.optimalRestoreTimeValue) {
				year = 0
			}
			year += 1
		}
		return ResponseEntity.ok(timeValues)
	}

	@GetMapping("/elements/{elementId}/relativeAge/{timeValue}")
	@Throws(InterruptedException::class, ExecutionException::class)
	fun getRelativeAge(
		@PathVariable elementId: String,
		@PathVariable timeValue: Double,
	): ResponseEntity<Double?> {
		val buildingPart = getBuildingPart(elementId)!!
		return ResponseEntity.ok(buildingPart.getRelativeAge(timeValue))
	}

	@GetMapping("/elements/{elementId}/nextRestoration/{ratingYear}/{condition}")
	@Throws(InterruptedException::class, ExecutionException::class)
	fun getNextRestoration(
		@PathVariable elementId: String,
		@PathVariable ratingYear: Int,
		@PathVariable condition: Double,
	): ResponseEntity<ProjectionPeriod?> {
		val buildingPart = getBuildingPart(elementId)!!
		return ResponseEntity.ok(buildingPart.getNextRestoration(1000000.0, ratingYear, condition))
	}

}
